File: src/hash/Azrael64.java
/* Copyright (C) 2007-2017 - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Jorge Medina <medinarosas.jorge@gmail.com>, December 2017
 */
package hash;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

import org.apache.commons.codec.binary.Hex;

/**
 * @author jmedina
 *
 */ 
public class Azrael64 implements FuncionHash {

	private static final boolean DEBUG_PARTIAL_HASH = false;
	private static final boolean DEBUG_INTERMIDIATE_HASH = false;

	private int rounds = 0;

	private long numIterations = 2;
	private int iteration = 0;
	private RandomAccessFile fileToPersist;

	private static final String EMPTY_STRING_1_IT =
			"d7ffe7b852fe0e97";

	private static final String EMPTY_STRING_2_IT = 
			"e99b67b4ddbf9272";

	/**
	 * 
	 */
	public Azrael64() {
		this(1);
	}
	
	/**
	 * 
	 */
	public Azrael64( long numIterations ) {
		this.numIterations = numIterations;
	}

	@Override
	public BigInteger getHash(byte[] input) {
		this.rounds  = 0;
		this.iteration  = 0;

		int promedioPro = 0;
		for( ; iteration<numIterations; ) {
			iteration++;
			input = getHashEval( input );
			
			String value = Hex.encodeHexString(input);
			
			if( "ffdb3d80fed96840".equals(value) ) {
				System.out.println( iteration +"="+value );
			}
			
			/*
			if( this.fileToPersist == null ) {
				try {
					String filename = "/home/jmedina/eclipse-workspace/HashUniforme/src/azrael64-500millions-hashes.txt";
					this.fileToPersist = new RandomAccessFile( new File(filename), "rw");
				} catch (FileNotFoundException e) {
					e.printStackTrace();
				}
			}
			try {
				this.fileToPersist.write( (value+"\n").getBytes(StandardCharsets.UTF_8) );
			} catch (IOException e) {
				e.printStackTrace();
			}*/
			
			
			if( DEBUG_INTERMIDIATE_HASH ) {
				
				//String value = Hex.encodeHexString(input);
				
				int min=256, max=-1;
				
				int promedioHex = 0;
				int byteNum = 0;
				for( int i=0; i<input.length; i++ ) {
					byteNum = input[i]<0? 127-input[i]: input[i];
					promedioHex += byteNum;
					
					if( byteNum < min ) {
						min = byteNum;
					}
					if( max < byteNum ) {
						max = byteNum;
					}
				}
				double promedio = promedioHex/(double)input.length;
				promedio = (promedio*100)/255;
				promedio = ((int)(promedio*1000000))/1000000d;
				
				StringBuffer sb = new StringBuffer();
				//sb.append( promedio );
				
				if( iteration > this.numIterations - 100 ) {
					promedioPro += promedio;
					
					sb.append( "[" )
					  .append( value )
					  .append( "]-[bits]=[" )
					  .append( input.length*8 )
					  .append( "]-[avg]=[" )
					  .append( promedio )
					  .append( "]-[min,max]=[" )
					  .append( min )
					  .append( "," )
					  .append( max )
					  .append( "]-[iter]=[" )
					  .append( iteration )
					  .append( "]" );					
					 
					//System.out.println( iteration +"---"+ sb.toString() );
					System.out.println( sb.toString() );
				}
			}
		}
		/*
		try {
			this.fileToPersist.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
		*/
		
		//System.out.println( promedioPro/100d );

		return new BigInteger( input );
	}

	public byte[] getHashEval( byte[] input ) {
		
		input = pad(input);

		long hash = 0;
		//long char1=0,char2=0,char3=0,char4=0,char5=0;

		// fractional part of sqrt(2) and sqrt(3)
		long IV1 = 0x6a09e667bb67ae85l;
		long IV2 = 0x3c6ef372a54ff53al;
		long IV3 = 0x510e527f9b05688cl;
		long IV4 = 0x1f83d9ab5be0cd19l;
		long IV5 = 0x428a2f9871374491l;
		long IV6 = 0xb5c0fbcfe9b5dba5l;
		long IV7 = 0x3956c25b59f111f1l;
		long IV8 = 0x923f82a4ab1c5ed5l;
		long IV9 = 0xd807aa9812835b01l;
		long IV10 = 0x243185be550c7dc3l;
		
		long char1=IV1;
		long char2=IV2;
		long char3=IV3;
		long char4=IV4;
		long char5=IV5;
		
		/*
		long sumaAnt1 = (long)input[ input.length-2 ];
		long sumaAnt2 = (long)input[ 2 ];
		long sumaAnt3 = 0;
		long sumaAnt4 = 0;
		long sumaAnt5 = 0;
		*/
		
		long sumaAnt1 = IV6;
		long sumaAnt2 = IV7;
		long sumaAnt3 = IV8;
		long sumaAnt4 = IV9;
		long sumaAnt5 = IV10;

		char1 += (long)input[ input.length-2 ];
		char2 += (long)input[ input.length-1 ];
		char3 += (long)input[ 0 ];
		char4 += (long)input[ 1 ];
		char5 += (long)input[ 2 ];
		sumaAnt5 += sumaAnt4;
		sumaAnt4 += sumaAnt3;
		sumaAnt3 += sumaAnt2;
		sumaAnt2 += sumaAnt1;
		sumaAnt1 += evaluaFuncBool( char1,char2,char3,char4,char5);

		// Main Loop.
		for( int i=1; i<input.length-1; i++ ) {
			char1 += sumaAnt1;
			char2 += char3;
			char3 += char4;
			char4 += (long)input[ i+1 ];
			char5 += sumaAnt2;
			sumaAnt5 += sumaAnt4;
			sumaAnt4 += sumaAnt3;
			sumaAnt3 += sumaAnt2;
			sumaAnt2 += sumaAnt1;
			sumaAnt1 += evaluaFuncBool( char1,char2,char3,char4,char5);
		}

		char1 += sumaAnt1;
		char2 += char3;
		char3 += char4;
		char4 += (long)input[ 0 ];
		char5 += sumaAnt2;
		sumaAnt5 += sumaAnt4;
		sumaAnt4 += sumaAnt3;
		sumaAnt3 += sumaAnt2;
		sumaAnt2 += sumaAnt1;
		sumaAnt1 += evaluaFuncBool( char1,char2,char3,char4,char5);
		
		if( DEBUG_PARTIAL_HASH ) {
			System.out.println( "**** END ACUMULACION 5x64: ("+rounds+") rounds" );
			System.out.println( "**** sumAnt5 = "+sumaAnt5 );
			System.out.println( "**** sumAnt4 = "+sumaAnt4 );
			System.out.println( "**** sumAnt3 = "+sumaAnt3 );
			System.out.println( "**** sumAnt2 = "+sumaAnt2 );
			System.out.println( "**** sumAnt1 = "+sumaAnt1 );
		}

		
		hash = ((sumaAnt1 << 48) & 0xffffffffffffffffL ) |
	               ((sumaAnt1+sumaAnt2 << 32) & 0xffffffffffffffffL ) |
	               ((sumaAnt1+sumaAnt2+sumaAnt3 << 16) & 0xffffffffL) |
	               ((sumaAnt3+sumaAnt4+sumaAnt5) & 0xffffffffL) +rounds;
					
		if( DEBUG_PARTIAL_HASH ) {
			System.out.println( "**** END APILACION 1x64:" );
			System.out.println( "**** hash = "+hash );
		}

		return longToBytes(hash);
	}
	
	public byte[] longToBytes(long x) {
	    ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
	    buffer.putLong(x);
	    return buffer.array();
	}

	private byte[] pad(byte[] data) {
        int length = data.length;
        int tail = length % 64;
        int padding;

        if ((64 - tail >= 9)) {
            padding = 64 - tail;
        } else {
            padding = 128 - tail;
        }

        byte[] pad = new byte[padding];
        pad[0] = (byte) 0x80;
        long bits = length * 8;
        for (int i = 0; i < 8; i++) {
            pad[pad.length - 1 - i] = (byte) ((bits >>> (8 * i)) & 0xFF);
        }

        byte[] output = new byte[length + padding];
        System.arraycopy(data, 0, output, 0, length);
        System.arraycopy(pad, 0, output, length, pad.length);

        return output;
    }
	
	public long evaluaFuncBool(Long char1, Long char2, Long char3, Long char4, Long char5) {
		rounds++;
		return (( char1 + char2 ) ^ ( char3 ^ char4 ) ^ char5) +
			   (( char1 & char2 ) ^ ( char3 + char4 ) ^ char5) +
			   (( char1 ^ char2 ) + ( char3 + char4 ) ^ char5) +
			   (( char1 ^ char2 ) ^ ( char3 + char4 ) ^ char5) +
			   (( char1 & char2 ) + ( char3 + char4 ) + char5) +
			   (( char1 & char2 ) + ( char3 + char4 ) ^ char5) +
			   (( char1 ^ char2 ) ^ ( char3 + char4 ) ^ char5) +
			   (( char1 | char2 ) ^ ( char3 + char4 ) ^ char5) +
			   (( char1 | char2 ) | ( char3 + char4 ) ^ char5) +
			   (( char1 + char2 ) + ( char3 + char4 ) ^ char5) +
			   (( char1 + char2 ) & ( char3 + char4 ) ^ char5) +
			   (( char1 ^ char2 ) + ( char3 ^ char4 ) ^ char5) +
			   (( char1 | char2 ) ^ ( char3 ^ char4 ) ^ char5) +
			   (( char1 + char2 ) + ( char3 ^ char4 ) ^ char5) +
			   (( char1 + char2 ) ^ ( char3 + char4 ) ^ char5) +
			   (( char1 + char2 ) & ( char3 + char4 ) + char5);
	}

	public String toString() {
		return "Azrael64 "+numIterations+"x";
	}
	
	public static void main2(String[] args) {
		
		long tope = 10000000000l;
		Azrael64 hash = new Azrael64(tope);
		
		long timeIni = System.currentTimeMillis();
		System.out.println( "=====> INIT ("+tope+")-"+timeIni );

		hash.getHash( "".getBytes(StandardCharsets.UTF_8) );
		
		long timeNow = System.currentTimeMillis() - timeIni;
		System.out.println( "TIME = "+
				"["+(timeNow/(1000.0))+"] secs,"+
				"["+(timeNow/(1000.0*60.0))+"] mins." );
	}

	/**
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		
		Azrael64 hash = new Azrael64(1);
		
		byte[] hash1 = hash.getHashEval( "".getBytes(StandardCharsets.UTF_8) );
		String hex1 = Hex.encodeHexString( hash1 );
		
		System.out.println( "1 => "+hex1 );
		
		byte[] hash2 = hash.getHashEval( hash1 );
		String hex2 = Hex.encodeHexString( hash2 );
		
		System.out.println( "2 => "+hex2 );
		
		System.out.println( "EMPTY 1 => "+(hex1.equals(EMPTY_STRING_1_IT)) );
		System.out.println( "EMPTY 2 => "+(hex2.equals(EMPTY_STRING_2_IT)) );
	}
}

File: src/c/azrael64.c
/* Copyright (C) 2007-2020 - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Jorge Medina <medinarosas.jorgealberto@gmail.com>, October 7th, 2020
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdint.h>
#include <inttypes.h>

static int rounds = 0;
static int iteration = 0;
static int iteraciones = 0;

#ifdef PRINT_HASH
  static const int numMostrar = 20;
#endif

static inline uint64_t evaluaFuncBool( register const uint64_t char1,
		register const uint64_t char2,
		register const uint64_t char3,
		register const uint64_t char4,
		register const uint64_t char5 ) {
	rounds++;
  	return ((( char1 + char2 ) ^ ( char3 ^ char4 )) ^ char5) +
	       ((( char1 & char2 ) ^ ( char3 + char4 ))	^ char5) +
	       ((( char1 ^ char2 ) + ( char3 + char4 )) ^ char5) +
	       ((( char1 ^ char2 ) ^ ( char3 + char4 )) ^ char5) +
	       ((( char1 & char2 ) + ( char3 + char4 )) + char5) +
	       ((( char1 & char2 ) + ( char3 + char4 )) ^ char5) +
	       ((( char1 ^ char2 ) ^ ( char3 + char4 )) ^ char5) +
	       ((( char1 | char2 ) ^ ( char3 + char4 )) ^ char5) +
	       (( char1 | char2 ) | (( char3 + char4 ) ^ char5)) +
	       ((( char1 + char2 ) + ( char3 + char4 )) ^ char5) +
	       ((( char1 + char2 ) & ( char3 + char4 )) ^ char5) +
	       ((( char1 ^ char2 ) + ( char3 ^ char4 )) ^ char5) +
	       ((( char1 | char2 ) ^ ( char3 ^ char4 )) ^ char5) +
	       ((( char1 + char2 ) + ( char3 ^ char4 )) ^ char5) +
	       ((( char1 + char2 ) ^ ( char3 + char4 )) ^ char5) +
	       (( char1 + char2 ) & (( char3 + char4 ) + char5));
 }

static inline char* pad( const char* data, int length, int padding, char* output ) {

  memcpy(output,data,length);

  int i;
  const uint64_t bits = length * 8;

  char pad[padding];
  pad[0] = (char) 0x80;
  for( i=1; i<=padding-1-8; i++ ) {
	 pad[i] = (char)0;
  }
  for( i=0; i < 8; i++ ) {
	 pad[padding - 1 - i] = (char) ((bits >> (8 * i)) & 0xFF);
  }
  memcpy(output+length,pad,padding);

  return output;
}

char* eval_hash( char* input, char* val, int inputLength ) {

  // Do padding.
  const int tail = inputLength % 64;
  int padding = 64 - tail >= 9? 64 - tail: 128 - tail;

  char output[inputLength+padding];
  input = pad(input,inputLength,padding,output);
  inputLength += padding;

  // Begin calculate..
  int i,j;
  uint64_t cha[5] = {0,0,0,0,0};
  uint64_t carrier[5] = {(uint64_t)input[inputLength-2],(uint64_t)input[2],0,0,0};
  register uint64_t hash1;

    // First iteration.
  cha[0] += carrier[0];
  cha[1] += (uint64_t)input[ inputLength-1 ];
  cha[2] += (uint64_t)input[ 0 ];
  cha[3] += (uint64_t)input[ 1 ];
  cha[4] += carrier[1];
  carrier[4] += carrier[3];
  carrier[3] += carrier[2];
  carrier[2] += carrier[1];
  carrier[1] += carrier[0];
  carrier[0] += evaluaFuncBool( cha[0],cha[1],cha[2],cha[3],cha[4]);
  
  // Main Loop.
  for( i=1; i<inputLength-1; i++ ) {
    cha[0] += carrier[0];
    cha[1] += cha[2];
    cha[2] += cha[3];
    cha[3] += (uint64_t)input[i+1];
    cha[4] += carrier[1];
    carrier[4] += carrier[3];
    carrier[3] += carrier[2];
    carrier[2] += carrier[1];
    carrier[1] += carrier[0];
    carrier[0] += evaluaFuncBool( cha[0],cha[1],cha[2],cha[3],cha[4]);
  }

  // Last iteration.
  cha[0] += carrier[0];
  cha[1] += cha[2];
  cha[2] += cha[3];
  cha[3] += (uint64_t)input[ 0 ];
  cha[4] += carrier[1];
  carrier[4] += carrier[3];
  carrier[3] += carrier[2];
  carrier[2] += carrier[1];
  carrier[1] += carrier[0];
  carrier[0] += evaluaFuncBool( cha[0],cha[1],cha[2],cha[3],cha[4]);

#ifdef DEBUG_PARTIAL_HASH
    printf("(1) END ACUMULACION (%d)\n", rounds );
    printf("carrier[4] = [%016" PRId64 "]!\n",carrier[4] );
    printf("carrier[3] = [%016" PRId64 "]!\n",carrier[3] );
    printf("carrier[2] = [%016" PRId64 "]!\n",carrier[2] );
    printf("carrier[1] = [%016" PRId64 "]!\n",carrier[1] );
    printf("carrier[0] = [%016" PRId64 "]!\n",carrier[0] );
#endif


  hash1 = ((carrier[0] << 48) & 0xffffffffffffffffL ) |
          (((carrier[0]+carrier[1]) << 32) & 0xffffffffffffffffL ) |
          (((carrier[0]+carrier[1]+carrier[2]) << 16) & 0xffffffffL) |
          ((carrier[2]+carrier[3]+carrier[4]) & 0xffffffffL);

#ifdef DEBUG_PARTIAL_HASH
    printf("(3) END APILACION (%d)\n", rounds );
    printf("hash[0] = [%016" PRId64 "] !\n",hash[0] );
#endif

  hash1 += rounds;

#ifdef PRINT_HASH
  if( iteration >= iteraciones - numMostrar ) {
    char hex[17];
    sprintf(hex,"%016" PRIx64,hash1 );
    printf( "%s\n", hex );
  }
#endif

  // Do output val.
  for(j=0; j<8; j++ ) {
    val[7-j] = hash1 >> 8*j;
  }

  return val;
}

static inline char* getHash( int numIterations, char* val1, char* val2, int size ) {

  iteraciones = numIterations;
  rounds  = 0;

  val1 = eval_hash( val1, val2, size );
  for( iteration = 1; iteration<numIterations; iteration++ ) {
    val1 = eval_hash( val1, val2, 8 );
  }

  return val2;
}

int main(int argc, char *argv[]) {

  int iter = 0;
  if( argc < 2 ) {
    printf( "Error: need one argument to hash!\n" );
    return 1;
  } else if( argc == 2 ) {
    iter = 1;
  } else {
    iter = atoi(argv[2]);
    if( iter <= 0 ) {
        printf( "Error: second argument need to be positive!\n" );
        return 1;
    }
  }

  char val2[64]; // output 64 bytes/512 bits.
  char* val1 = argv[1]; // input
  int size = strlen(val1);

  getHash( iter, val1, val2, size );

  return 0;
}
